fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1428,22 +1471,70 @@ def forward(self, *args: Any, **kwargs: Any) -> torch.Tensor:
+        is_bf16 = self.compute_dtype in [
+            torch.bfloat16,
+            torch.float8_e5m2,
+            torch.float8_e4m3fn,
+        ]
         if self._is_root and self.mixed_precision:
             args, kwargs = cast_floats_to_right_precision(True, True, is_bf16, *args, **kwargs)
 
+        just_added_to_fsdp_forward_ordering = False
         if self not in self._fsdp_forward_ordering:
             self._my_fsdp_instance_idx = len(self._fsdp_forward_ordering)
             self._fsdp_forward_ordering.append(self)
+            just_added_to_fsdp_forward_ordering = True
 
         # If enabled, convert the input to FP32 if we are in full precision.
         # no_grad is not used because the input might be for a non-root instance,
         # which mean autograd needs to go through the conversion.
         if self.force_input_to_fp32 and not self.mixed_precision:
             args, kwargs = cast_floats_to_right_precision(False, False, is_bf16, *args, **kwargs)
 
+        # need to use fp32_to_fp16 stream since _cast_fp32_param_shards_to_fp16
+        # depends on this block.
+        with torch.no_grad(), torch.cuda.stream(self._streams["fp32_to_fp16"]):
+            # Collect parameters to update scale/scale_inv before we
+            # _cast_fp32_param_shards_to_fp16 that uses fp8 scale to quantize
+            # before all-gather.
+            # These include params we prefetch all-gather.
+            params = []
+            if self._my_fsdp_instance_idx < len(self._fsdp_forward_ordering) - 1:
+                if self._my_fsdp_instance_idx == 0 and self._is_fp8_dtype():
+                    # The first FSDP instance didn't have chance to prefetch
+                    params = self.params
+                if self._fsdp_forward_ordering[self._my_fsdp_instance_idx + 1]._is_fp8_dtype():
+                    # FSDP instance we'll prefetch all-gather
+                    params.extend(self._fsdp_forward_ordering[self._my_fsdp_instance_idx + 1].params)
+            elif just_added_to_fsdp_forward_ordering:
+                # In the first iteration, we didn't have chance to record
+                # fsdp_instance_idx to prefetch
+                if self._is_fp8_dtype():
+                    params = self.params
+
+            for p in params:
+                if not isinstance(p, FlatParameter):
+                    continue
+                d = {info[0]: info[1] for info in p._param_infos}
+                for n, m in d.items():
+                    # Previous iteration was grad_enabled
+                    if m.fp8_meta.get("update_amax_and_scale_fwd", False):
+                        if m.fp8_meta["recipe"].reduce_amax:
+                            FP8GlobalStateManager.copy_amax_from_global_buffer(
+                                m.fp8_meta, forward=True
+                            )
+                            # FIXME update_weight_scale_inv is only True for the first micro-batch
```

Comment:
As discussed it's OK to update weight_scale_inv multiple times but it can be still annoying if we see numerical differences btw no-PP and PP (actually just micro-batching). Actually I wonder we can check is_first_microbatch in kwargs to skip this.